pageserver/src/object_repository.rs

```diff
@@ -0,0 +1,909 @@
+
+                                        truncated += 1;
+                                    } else {
+                                        trace!(
+                                            "Keeping horizon page {} blk {} at {}",
+                                            key.tag.rel,
+                                            key.tag.blknum,
+                                            key.lsn
+                                        );
+                                    }
+                                }
+                            } else {
+                                trace!(
+                                    "Last page {} blk {} at {}, horizon {}",
+                                    key.tag.rel,
+                                    key.tag.blknum,
+                                    key.lsn,
+                                    horizon
+                                );
+                            }
+                            // remove records prior to horizon
+                            loop {
+                                iter.prev();
+                                if !iter.valid() {
+                                    break;
+                                }
+                                let key = StorageKey::des(iter.key().unwrap());
+                                if key.tag != maxkey.tag {
+                                    break;
+                                }
+                                let mut val = StorageValue::des(iter.value().unwrap());
+                                if val.alive {
+                                    val.alive = false;
+                                    self.storage.put(key, val)?;
+                                    deleted += 1;
+                                    trace!(
+                                        "deleted: {} blk {} at {}",
+                                        key.tag.rel,
+                                        key.tag.blknum,
+                                        key.lsn
+                                    );
+                                } else {
+                                    break;
+                                }
+                            }
+                        }
+                        maxkey = minkey;
+                    } else {
+                        break;
+                    }
+                }
+                info!("Garbage collection completed in {:?}:\n{} version chains inspected, {} pages reconstructed, {} version histories truncated, {} versions deleted",
+					  now.elapsed(), inspected, reconstructed, truncated, deleted);
+            }
+             */
+        }
+    }
+
+    //
+    // Wait until WAL has been received up to the given LSN.
+    //
+    fn wait_lsn(&self, mut lsn: Lsn) -> Result<Lsn> {
+        // When invalid LSN is requested, it means "don't wait, return latest version of the page"
+        // This is necessary for bootstrap.
+        if lsn == Lsn(0) {
```

Comment:
It seems weird to me to have a "don't wait" option hiding inside a function called "wait".

I don't love use of "magic" values to trigger alternate behavior.

I'm starting to wonder if there are other questionable uses `Lsn(0)`. Many of them are used as a hypothetical `Lsn::MIN` but others seem to be indicating one flavor or another of "no such value", which seems wrong to me.

Reply:
> It seems weird to me to have a "don't wait" option hiding inside a function called "wait".
> 
> I don't love use of "magic" values to trigger alternate behavior.

Agreed. I don't think we should need this magic nowait behavior at all, actually. Some discussion on that at https://github.com/zenithdb/zenith/pull/175#issuecomment-846227525.

This is a pre-existing issue, though, I just moved this code as it is. Let's address that as a separate PR.

> I'm starting to wonder if there are other questionable uses `Lsn(0)`. Many of them are used as a hypothetical `Lsn::MIN` but others seem to be indicating one flavor or another of "no such value", which seems wrong to me.

Yeah. `Lsn::MIN` and `Lsn::MAX` would be at least make the usage in the search keys more clear.